drivers/syslog/syslog_channel.c

```diff
@@ -135,14 +139,82 @@ static int syslog_default_putc(int ch)
 
 int syslog_channel(FAR const struct syslog_channel_s *channel)
 {
+#if (CONFIG_SYSLOG_MAX_CHANNELS != 1)
+  int i;
+#endif
+
   DEBUGASSERT(channel != NULL);
 
   if (channel != NULL)
     {
       DEBUGASSERT(channel->sc_putc != NULL && channel->sc_force != NULL);
 
-      g_syslog_channel = channel;
+#if (CONFIG_SYSLOG_MAX_CHANNELS == 1)
+      g_syslog_channel[0] = channel;
       return OK;
+#else
+      for (i = 0; i < CONFIG_SYSLOG_MAX_CHANNELS; i++)
+        {
+          if (g_syslog_channel[i] == NULL)
+            {
+              g_syslog_channel[i] = channel;
+              return OK;
+            }
+          else if (g_syslog_channel[i] == channel)
+            {
+              return OK;
+            }
+        }
+#endif
+    }
+
+  return -EINVAL;
+}
+
+/****************************************************************************
+ * Name: syslog_channel_remove
+ *
+ * Description:
+ *   Removes an already configured SYSLOG channel from the list of used
+ *   channels.
+ *
+ * Input Parameters:
+ *   channel - Provides the interface to the channel to be removed.
+ *
+ * Returned Value:
+ *   Zero (OK) is returned on success.  A negated errno value is returned
+ *   on any failure.
+ *
+ ****************************************************************************/
+
+int syslog_channel_remove(FAR const struct syslog_channel_s *channel)
```

Comment:
There are more work to support the add/remove channel dynamically:

1. All methods need be protected by spin lock or critical section.
2. Need notify syslog driver it has been removed

So, it's may better to split the work into two part:

1. Support register multiple channels only in the initialization phase
2. Support register/unregister the channel in the runtime dynamically

Reply:
> All methods need be protected by spin lock or critical section.

I just cannot understand how syslog is protected now.  
Following the code from `syslog()` to `nx_vsyslog()`, I don't see any protection (e.g. semaphore etc).

Shouldn't syslog be thread-safe? In the sense that multiple tasks may log simultaneously?

Reply:
> > All methods need be protected by spin lock or critical section.
> 
> I just cannot understand how syslog is protected now.
> Following the code from `syslog()` to `nx_vsyslog()`, I don't see any protection (e.g. semaphore etc).
> 

The protection doesn't need before since g_syslog_channel is never changed at the runtime.

> Shouldn't syslog be thread-safe? In the sense that multiple tasks may log simultaneously?

Yes, syslog is not only thead-safe but also interrupt safe(true for uart, ram and lwl, false for file and dev). The protection is done inside the driver level. syslog/nx_vsyslog use stack variable only except g_syslog_channel, which need to protect now with your new code.

Reply:
But, till `nx_vsyslog()` there is no protection, which mean that this function may be called concurrently by multiple threads.

Since `nx_vsyslog()` makes several (unprotected) calls to `lib_sprintf()`, I would expect that the output can be garbled.

Something like:
```
[1616508998.000][1616508998.001] message2 message1
```
instead of:
```
[1616508998.000] message1
[1616508998.001] message2
```

Reply:
> But, till `nx_vsyslog()` there is no protection, which mean that this function may be called concurrently by multiple threads.
> 

Yes, of course.

> Since `nx_vsyslog()` makes several (unprotected) calls to `lib_sprintf()`, I would expect that the output can be garbled.
> 

lib_sprintf output to a local buffer, different thread has the different buffer, so no problem here.

> Something like:
> 
> ```
> [1616508998.000][1616508998.001] message2 message1
> ```
> 
> instead of:
> 
> ```
> [1616508998.000] message1
> [1616508998.001] message2
> ```

Is it the real output? or just your imagination.

Reply:
No this is not real output.  
But still I cannot understand this.

`lib_sprintf()` is finally calling the `putc(c,stream)` macro, which may end up to a plain call to `syslog_putc()`.

I don't see any buffering take place, that will concatenate all `lib_sprintf()` calls to a single string to be outputed atomically.

Reply:
There are two stream(emerge v.s. syslog):
```
  if (priority == LOG_EMERG)
    {
      /* Use the SYSLOG emergency stream */

      emergstream(&stream.public);
    }
  else
    {
      /* Use the normal SYSLOG stream */

      syslogstream_create(&stream);
    }
```
The first(emerge) call syslog_force directly, the second(syslog) save the result into iob buffer and flush through syslog_write at the end.

Reply:
I am sorry but I cannot see what you describe...

Here is a stack trace while the logger is outputing characters:
![Screenshot from 2021-03-23 17-08-28](https://user-images.githubusercontent.com/46975045/112172814-b2578900-8bfd-11eb-8249-42b1f00ed222.png)

I don't see in any of these functions any kind of buffering.

---

In fact is was very easy to replicate what I imagined above.  
I did:
```c
	task_create("test1", 128, 2048, test_th, 0);
	task_create("test2", 129, 2048, test_th, 0);
```

and
```c
int test_th(int argc, char ** argv)
{
	while(1)
	{
		usleep(rand() % 10000);
		syslog(LOG_INFO, "A test message.\n");
	}
}
```

And I got this awesome output:

![image](https://user-images.githubusercontent.com/46975045/112173149-fd719c00-8bfd-11eb-8c10-d367adf26592.png)

Reply:
Ok, you forget to enable CONFIG_SYSLOG_BUFFER, please try it.

Reply:
Yes, I haven't enabled it.
This option is by default disabled and unavailable.

It depends on `CONFIG_SYSLOG_WRITE`, which is not avaiable in the default NSH configuration.  
I had it selected by enabling `CONFIG_SYSLOG_FILE` (which is unrelated and unneeded in this example).

Interestingly enough, the issue persists  (I would expect the output to be OK now, too).  
I haven't managed to follow the code, and how this is caused.

In any case though, I would expect the system logger to be thread-safe under all circumstances...  
I would insist that a semaphore has to be added to all the public functions of the syslog.

Reply:
I am not sure if this is now the issue, but the buffered syslog output is surely to have problem too.

If the output string is longer than `CONFIG_IOB_BUFSIZE`, then the buffer will be outputed prematurely, causing the exact same race condition as without buffer.

The fact that the buffer may be flushed before the whole syslog message has been written to it, defeats completely its puprose...

Reply:
> Interestingly enough, the issue persists (I would expect the output to be OK now, too).
> I haven't managed to follow the code, and how this is caused.
> 
> In any case though, I would expect the system logger to be thread-safe under all circumstances...

it depends on how you define thread-safe. The default behaviour doesn't cause the hang or lose the data.

> I would insist that a semaphore has to be added to all the public functions of the syslog.

No, you can't use semaphore here, syslog can be called from interrupt context.

> I am not sure if this is now the issue, but the buffered syslog output is surely to have problem too.
> 
> If the output string is longer than `CONFIG_IOB_BUFSIZE`, then the buffer will be outputed prematurely, causing the exact same race condition as without buffer.
> 
> The fact that the buffer may be flushed before the whole syslog message has been written to it, defeats completely its puprose...

Yes, if your output is longer than CONFIG_IOB_BUFSIZE, the buffer has to be flushed in the middle.